tests/unit/test_sender.py

```diff
@@ -0,0 +1,185 @@
+        # Check metric has been associated with correct online run
+        _run_metric = next(_online_run.metrics)
+        assert _run_metric[0] == 'x'
+        assert _run_metric[1]["count"] == 1
+        assert _run_metric[1]["min"] == i
+    
+        # Create a new offline metric with offline run ID
+        _metrics = Metrics.new(
+            run=_offline_run_id,
+            metrics=[
+                {
+                    "timestamp": datetime.datetime.now().strftime(DATETIME_FORMAT),
+                    "time": 2,
+                    "step": 2,
+                    "values": {"x": 2},
+                }
+            ],
+            offline=True
+        )
+        _metrics.commit()
+    
+    # Run sender again, check online ID is correctly loaded from file and substituted for offline ID
+    with caplog.at_level(logging.ERROR):
+        sender(threading_threshold=1 if parallel else 10)
+    
+    assert not caplog.text
+    
+    # Check metric uploaded correctly
+    for _online_run in _online_runs:
+        _online_run.refresh()
+        _run_metric = next(_online_run.metrics)
+        assert _run_metric[0] == 'x'
+        assert _run_metric[1]["count"] == 2
+        
+    # Check all files for runs and metrics deleted once they were processed
+    assert len(list(pathlib.Path(offline_cache_setup.name).joinpath("runs").iterdir())) == 0
+    assert len(list(pathlib.Path(offline_cache_setup.name).joinpath("metrics").iterdir())) == 0
+
+@pytest.mark.parametrize("parallel", (True, False))
+def test_send_heartbeat(offline_cache_setup, parallel, mocker):
+    # Create an offline run
+    _uuid: str = f"{uuid.uuid4()}".split("-")[0]
+    _path = f"/simvue_unit_testing/objects/folder/{_uuid}"
+    _folder = Folder.new(path=_path, offline=True)
+    _folder.commit()
+    
+    _offline_runs = []
+    
+    for i in range(5):
+        _name = f"test_sender_server_ids-{_uuid}-{i}"
+        _run = Run.new(name=_name, folder=_path, offline=True, heartbeat_timeout=1, status="running")
+        _run.commit()
+        
+        _offline_runs.append(_run)
+    
+    _id_mapping = sender(threading_threshold=1 if parallel else 10)
+    _online_runs = [Run(identifier=_id_mapping.get(_offline_run.id)) for _offline_run in _offline_runs]
+    assert all([_online_run.status == "running" for _online_run in _online_runs])
+    
+    spy_put = mocker.spy(requests, "put")
+    
+    # Create heartbeat and send every 0.5s for 5s
+    for i in range(10):
+        time.sleep(0.5)
+        [_offline_run.send_heartbeat() for _offline_run in _offline_runs]
+        sender(threading_threshold=1 if parallel else 10)
+        
+    # Check requests.put() endpoint called 50 times - once for each of the 5 runs, on all 10 iterations
+    assert spy_put.call_count == 50
+        
+    # Get online runs and check all running
+    [_online_run.refresh() for _online_run in _online_runs]
+    assert all([_online_run.status == "running" for _online_run in _online_runs])
```

Comment:
~~Are you using the pre-commit hooks? Just there is a missing new line, I thought Ruff added this~~
Linting doesn't run on tests